### PR TITLE
New net-p2p/dogecoin package

### PR DESCRIPTION
**Pull request by victorsk2019**
Signed-off-by: Victor Skovorodnikov <victor3.14@yandex.com>

Suggested addition of net-p2p/dogecoin package as per [Bugzilla #811492](https://bugs.gentoo.org/811492) primarily for Dogecoin developers and testers.

### CODE REVIEW

**Comment by victorsk2019** [comment]
Signed-off-by: Victor Skovorodnikov <victor3.14@yandex.com>

File: net-p2p/dogecoin/dogecoin-1.14.4.ebuild [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Copyright 2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DESCRIPTION="Dogecoin (unstable) 1.14.4 Development version with new code for future release."
```
Comment (AbcSxyZ) [inline net-p2p/dogecoin/dogecoin-1.14.4.ebuild:6]: To add to https://bugs.gentoo.org/811492 comments :
> Please be more specific in the DESCRIPTION. The objective description may be <=80 characters long.

To be more explicit about the description, I can suggest  : *"Dogecoin Core ${PV} development version of blockchain node and RPC server."*

You can inspire yourself from the introduction sentence of the repository, it's going to be rewrite : https://github.com/dogecoin/dogecoin/pull/2570

File: net-p2p/dogecoin/metadata.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	<maintainer type="person" proxied="yes">
+		<email>victor3.14@yandex.com</email>
+		<name>Victor Skovorodnikov</name>
+	</maintainer>
+	<maintainer type="project" proxied="proxy">
+		<email>proxy-maint@gentoo.org</email>
+		<name>Proxy Maintainers</name>
+	</maintainer>
+	<use>
+		<flag name="gui">Build Dogecoin with dogecoin-qt GUI support</flag>
```
Comment (AbcSxyZ) [inline net-p2p/dogecoin/metadata.xml:18]: The real name of the software is Dogecoin Core, not Dogecoin. Maybe you should also speak about `dogecoind` ?

**Comment by patricklodder** [comment]
Re: a844eed - created builds for development versions...

Please note that running any of these versions can lead to loss of funds - do you really want to expose those?

**Comment by victorsk2019** [comment]
> Re: [a844eed](https://github.com/gentoo/gentoo/commit/a844eed16f370c1696b5601d3ac0970ab20c65a3) - created builds for development versions...
> 
> Please note that running any of these versions can lead to loss of funds - do you really want to expose those?

How different from them being  publicly offered on GitHub?

**Comment by patricklodder** [comment]
> How different from them being publicly offered on GitHub?

It's different because you're making it more generally available through re-distribution, I think. But we can discuss that on a case by case basis:

- I'm sure that you know that 1.17 has not been worked on for 2 years and has never made it to alpha.
- 1.18 doesn't have a single Dogecoin commit so I doubt you tested it. 

Both are also outdated codebases and unmaintained, so I am not sure why you would re-distribution. Anyone developing either of these branches is imho a waste of time, but feel free to disagree with that. 

- 1.21 in turn... doesn't sync yet. You probably know that?

These type of branches are perhaps better to be distributed after an alpha release rather than before that.

- 1.14.6 is a development branch that I literally opened this Monday... but that's the only one where you could have a case for running on edge, because at least that one had a lot of testing of 99% of the code. It's still risky though, so I just thought to warn you that that risk has to be communicated clearly.

I can see the benefits of making edge code available on top of tagged releases, and I think that it's a good thing that you're doing this. However, I feel that just exposing everything that has a higher version number than the latest release should get some reconsideration.

**Comment by victorsk2019** [comment]
First of, I am not sure why you expose unmaintained and outdated versions (particularly with higher version numbers) on GitHub - it obviously creates some confusion.  In fact, the very reason why I include these different versions is because I assumed they are being worked on at various degrees/stages and areas of progress/enhancements (as their different versions may indicate) (otherwise why would you have them exposed in the first place?)

> I am not sure why you would re-distribute

I don't think I am re-distributing anything, really.  The ebuild script is pulling packages from the same Dogecoin Core repository.  I am merely offering a convenient way for developers and testers to build these packages that are  located in that repository using Gentoo's package management tools (as I mention  this in metadata.xml, actually).

So, again, my intent was to provide a convenient way for developers and testers to build  various in-development Dogecoin Core versions.  I was obviously not aware that most of these development versions are in the state they are in.  

I have closed this PR and will re-open it once I re-work it with different package contents.
  
Thank you for your feedback  and I would recommend you remove those obsolete/deprecated/defunct development versions of which you mention.

**Comment by patricklodder** [comment]
> otherwise why would you have them exposed in the first place?

- 1.17 and 1.18: Because the sole developer that created these branches has requested these to be excluded from my deletion spree when I deleted everything else for exactly this reason. I share your concern, there's just not much I can do about it.
- 1.21: We need a public place to work on big refactoring projects... this is one of them. It needs to be public so that it can be collaborated on. Exposing it through a package manager before it has seen any meaningful QA is imho very risky.
- 1.14.6: This is the staging tree for the next maintenance release. This also simply needs to be public for it to be developed.

> So, again, my intent was to provide a convenient way for developers and testers to build various in-development Dogecoin Core versions.

Alright! I guess that's a goal on its own! 

I'm left with... why not package and distribute stable versions? Wouldn't that help a broader audience of gentoo users?

**Comment by thesamesam** [comment]
FWIW, I would not be comfortable as a Gentoo developer merging a PR which exposed development versions by default when someone tries to install Dogecoin.

It would be OK to provide them 'masked' (which means users have to quite aggressively opt in to it) though, if upstream (@patricklodder) were comfortable with that.

But generally I'm biased against adding unstable versions at all. Happy to consider a masked approach or something similar if they're really useful for folks though!

Thanks for engaging to both of you btw. Let me know if I can help here.

**Comment by victorsk2019** [comment]
> why not package and distribute stable versions? Wouldn't that help a broader audience of gentoo users?

Yes, I will work on this instead, thank you.

**Comment by patricklodder** [comment]
> It would be OK to provide them 'masked' (which means users have to quite aggressively opt in to it) though, if upstream were comfortable with that.

I think that this (masking) could be okay for alpha/beta/rc tags, maybe also minor release branches but I don't see an immediate benefit. I'd still be extremely hesitant to do it on unmaintained and/or immature branches that have not yet seen any testing. For me personally, as a former Gentoo user, messing around with masks has probably given me more interesting hours with a broken system than is healthy, and although I'm sure it's much more mature now than back in the early days, I do wonder if it's worth the trouble.

**Comment by victorsk2019** [comment]
Hello!  I need some clarification on the idea of exposing stable Dogecoin Core version to a broader audience.  The reason why I focused on ebuilds for unstable test developer version is because, in my mind, it would protect me from any potential blame for issues with the software.  It is my understanding that those who would work with unstable developer version would know what they are doing, understand the risks that the versions they are working with are "rough around the edges" and are for test purposes only and so users of these versions would not be putting funds they can't afford to risk loosing.  When Patrick mentions that I am "re-distributing" Dogecoin Core in these packages, this appears to make me responsible for the behaviour of the software and it is exactly why I chose to provide versions for a subset of advanced/expert users who would be less likely to go after me if something goes wrong.  Notwithstanding the fact that obtaining this software is no different from using 'wget' or 'git clone' commands by pulling sources from main repository.  And it would also be great to have Dogecoin Core to be a part of Gentoo's and (I hope) Calculate Linux packages.

There seem to be stable versions being provided for Litecoin and Bitcoin but my main concern is if I provide ebuilds for stable Dogecoin Core version for broader audience, would I be held responsible for any issues with Dogecoin Core software?  I just don't want any problems 😎  Thanks!

**Comment by victorsk2019** [comment]
Hi,

@patricklodder could there ever be a case when wallet is not a part of dogecoin-qt installation?  I, personally, don't think so and hence 'wallet' as USE flag may not be necessary for dogecoin-qt package?  It makes sense for dogecoind package though.

Thanks,
Victor.

**Comment by patricklodder** [comment]
> could there ever be a case when wallet is not a part of dogecoin-qt installation?

Yes. You can compile without wallet and get just a graphical interface for statistics, console interaction and peer management. I personally don't run it like that, but some people do.

**Comment by victorsk2019** [comment]
Thank you 👍

**Comment by victorsk2019** [comment]
@patricklodder  sorry, one more question, there's a configure option: `--with-intel-avx2  Build with intel avx2 (default is no)`.  My understanding is Avx2 instruction set optimizes mathematical/cryptographic operations, would it then not be helpful for Dogecoin synchronization process?  Of course, the advantage of Gentoo is you can easily fine-tune builds to your CPU architecture but should 'avx2' be part of USE flag option for Dogecoin build?

**Comment by patricklodder** [comment]
> Of course, the advantage of Gentoo is you can easily fine-tune builds to your CPU architecture but should 'avx2' be part of USE flag option for Dogecoin build?

It's experimental. If you choose to expose this, I think it should be marked as such. Note that it requires additional dependencies.

**Comment by victorsk2019** [comment]
Hi,

I've uploaded new dogecoin-qt and dogecoind packages into my branch: 

https://github.com/victorsk2019/gentoo/tree/master/net-p2p

I try keeping it simple for starters and will work on adding 'avx2' dependency later on because it seems like a cool build option to have.  Please feel free to go over these packages and let me know if there are any problems or if I can submit a PR for them.

Thanks,
Victor.

**Comment by patricklodder** [comment]
I have commented on your last commit in there. Overall looks good, one bigger issue with bdb, otherwise just some minor things

**Comment by victorsk2019** [comment]
Hi @thesamesam @patricklodder 

I am trying to submit a PR for net-p2p/dogecoind and net-p2p/dogecoin-qt packages https://github.com/gentoo/gentoo/pull/23449 but the PR also has records of closed net-p2p/dogecoin package.  I wouldn't want to mix commit information with closed net-p2p/dogecoin package in my PR.  Is there a way to only have net-p2p/dogecoind and net-p2p/dogecoin-qt packages showing in the PR?  It would probably be easier for reviewer to only see new packages in the PR?  Sorry, I don't do this very often :/

**Comment by juippis** [comment]
You might want to read up on `git rebase` - that's what you're looking for here.

As to packaging dogecoin in ::gentoo, I don't think there's currently much interest since even our bitcoin packages are a bit messy. And cryptocurrency is a fragile subject, so we'd really need to be able to provide quality support for these, and react fast. I don't think proxy-maintainers are **currently** up for that.

**Comment by paulkass** [comment]
Curious if by any chance there has been any progress on adding an ebuild of `dogecoin` to either this repository or some other overlay?

**Comment by paulkass** [comment]
Thanks for the swift reply @victorsk2019 . It's unfortunate that the official gentoo repo won't accept it at this time, however, it would also be a shame to let all of this work go to waste. Perhaps you can upload these ebuilds to the [GURU](https://wiki.gentoo.org/wiki/Project:GURU) repository? It's much more accommodating to unstable packages.

**Pull request closed without merging**